rascal-vscode-extension/src/RascalFileSystemProviders.ts

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (c) 2018-2021, NWO-I CWI and Swat.engineering
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+import { create } from 'node:domain';
+import { normalize } from 'node:path';
+import * as vscode from 'vscode';
+import { TextEncoder } from 'util';
+import {LanguageClient, LanguageClientOptions, ServerOptions, StreamInfo, integer } from 'vscode-languageclient/node';
+
+export class RascalFileSystemProvider implements vscode.FileSystemProvider {
+    readonly client: LanguageClient;
+    private readonly _emitter = new vscode.EventEmitter<vscode.FileChangeEvent[]>();
+    readonly onDidChangeFile: vscode.Event<vscode.FileChangeEvent[]> = this._emitter.event;
+    private readonly protectedSchemes:string[] = ["file", "http", "https", "unknown"];
+
+    /**
+     * Registers a single FileSystemProvider for every URI scheme that Rascal supports, except
+     * for file, http and https.
+     *
+     * @param client to use as a server for the file system provider methods
+     */
+    constructor (client:LanguageClient) {
+        this.client = client;
+
+        client.onNotification("rascal/filesystem/onDidChangeFile", (event:vscode.FileChangeEvent) => {
+            this._emitter.fire([event]);
```

Comment:
Why is this happening? what does it do?

Reply:
this is when a folder is being watched and the watch implementation reports that a change has happened. The listeners inside vscode expect a FileChangeEvent to be triggered.

Reply:
so it's basically passing on the event from the Java based part back into the TypeScript world.